File: Sources/LFGKit/CorruptionSignature.swift
import Foundation

// MARK: - ShadowOutcome

/// Outcome of a readonly shadow-attach probe used during Class B/C discrimination.
///
/// Shadow attach is only attempted when the primary attach failed (`hdiutil_attach_errno != nil`).
/// When the primary attach succeeded (errno nil), use `.notAttempted`.
public enum ShadowOutcome: Codable, Equatable, Sendable {

    /// Shadow attach was not attempted (primary attach succeeded or Class A pre-flight cut in).
    case notAttempted

    /// Shadow attach returned a device node; `fsck_apfs -yn` may or may not have found errors.
    ///
    /// - Parameter fsckClean: `true` when `fsck_apfs -yn` exited 0 (no errors found).
    case success(fsckClean: Bool)

    /// Shadow attach failed with the given reason string (e.g. "Resource temporarily unavailable").
    case failure(reason: String)

    // MARK: Codable

    private enum CodingKeys: String, CodingKey {
        case kind
        case fsckClean = "fsck_clean"
        case reason
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let kind = try c.decode(String.self, forKey: .kind)
        switch kind {
        case "not_attempted":
            self = .notAttempted
        case "success":
            let clean = try c.decodeIfPresent(Bool.self, forKey: .fsckClean) ?? true
            self = .success(fsckClean: clean)
        case "failure":
            let reason = try c.decodeIfPresent(String.self, forKey: .reason) ?? ""
            self = .failure(reason: reason)
        default:
            self = .notAttempted
        }
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .notAttempted:
            try c.encode("not_attempted", forKey: .kind)
        case .success(let fsckClean):
            try c.encode("success", forKey: .kind)
            try c.encode(fsckClean, forKey: .fsckClean)
        case .failure(let reason):
            try c.encode("failure", forKey: .kind)
            try c.encode(reason, forKey: .reason)
        }
    }
}

// MARK: - CorruptionSignature

/// Three-tuple signal used by `CorruptionClassifier` to assign a `CorruptionClass`.
///
/// Mirrors the `(hdiutil_attach_errno, container_visible, shadow_attach_outcome)` taxonomy
/// from `corruption-history-taxonomy.md` §4 (LFG-93).
///
/// Usage:
/// ```swift
/// // Class A — stale half-attach
/// let sig = CorruptionSignature(
///     hdiutilErrno: EBUSY,
///     containerVisible: false,
///     shadowAttachOutcome: .notAttempted
/// )!
/// let cls = CorruptionClassifier.classify(sig)  // → .classA
/// ```
public struct CorruptionSignature: Codable, Equatable, Sendable {

    // MARK: Properties

    /// The errno returned by a failed `hdiutil attach` attempt, or `nil` when attach succeeded.
    ///
    /// Canonical values from the corpus:
    /// - `nil`   — attach returned exit 0 (success; container/mount presence still unconfirmed)
    /// - `EBUSY` (16) — image already attached to a stale device node (Class A indicator)
    /// - `EIO`   (5)  — superblock invalid / errno 5 at block 0 (Class B indicator)
    public let hdiutilErrno: Int32?

    /// Whether `diskutil list /dev/diskN` showed an APFS container under the device node
    /// obtained after attach (or from `hdiutil info` in the pre-flight check).
    ///
    /// `false` when no device node was obtained at all (attach failed before device
    /// reservation). `false` in Class A (device present but no container under it).
    /// `true` in Class B/C/D/E (container is visible after attach).
    public let containerVisible: Bool

    /// Outcome of the shadow-attach probe, if it was attempted.
    ///
    /// Only relevant when `hdiutilErrno != nil` (primary attach failed). For Class A
    /// the pre-flight cuts in before the shadow probe, so this is `.notAttempted`.
    public let shadowAttachOutcome: ShadowOutcome

    // MARK: Failable init

    /// Creates a `CorruptionSignature` with the given tuple values.
    ///
    /// Returns `nil` when the combination is internally inconsistent:
    /// - Shadow attach succeeded (`success`) requires a prior failed primary attach
    ///   (`hdiutilErrno != nil`). A succeeded primary attach + shadow success is contradictory.
    ///
    /// - Parameters:
    ///   - hdiutilErrno: The errno from a failed attach, or `nil` on success.
    ///   - containerVisible: Whether an APFS container was observed under the device node.
    ///   - shadowAttachOutcome: The shadow-probe result.
    public init?(
        hdiutilErrno: Int32?,
        containerVisible: Bool,
        shadowAttachOutcome: ShadowOutcome
    ) {
        // Shadow success implies primary attach failed.
        if case .success = shadowAttachOutcome, hdiutilErrno == nil {
            return nil
        }
        self.hdiutilErrno = hdiutilErrno
        self.containerVisible = containerVisible
        self.shadowAttachOutcome = shadowAttachOutcome
    }

    // MARK: Codable

    private enum CodingKeys: String, CodingKey {
        case hdiutilErrno       = "hdiutil_errno"
        case containerVisible   = "container_visible"
        case shadowAttachOutcome = "shadow_attach_outcome"
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        hdiutilErrno       = try c.decodeIfPresent(Int32.self, forKey: .hdiutilErrno)
        containerVisible   = try c.decode(Bool.self, forKey: .containerVisible)
        shadowAttachOutcome = try c.decode(ShadowOutcome.self, forKey: .shadowAttachOutcome)
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encodeIfPresent(hdiutilErrno, forKey: .hdiutilErrno)
        try c.encode(containerVisible, forKey: .containerVisible)
        try c.encode(shadowAttachOutcome, forKey: .shadowAttachOutcome)
    }
}

File: devdrive/corruption_signature.py
"""CorruptionSignature — Python mirror of the Swift CorruptionSignature struct.

Carries the 3-tuple ``(hdiutil_errno, container_visible, shadow_attach_outcome)``
that is the canonical classification input for all three language runtimes per
ADR-003. Matches the JSON wire format consumed by
``POST /api/devdrive/backends/{backend_id}/state`` (LFG-98 §4.4).

Field names use snake_case to match the Phoenix evidence schema.
``ShadowOutcome`` encodes identically to the Elixir/Swift counterparts per
ADR-003 §Decision paragraph 4.

    not_attempted  — primary attach succeeded, or Class A pre-flight fired
    success        — shadow attached; fsck_apfs -yn clean flag carried
    failure        — shadow attach failed

Typical usage::

    from devdrive.corruption_signature import CorruptionSignature, ShadowOutcome

    sig = CorruptionSignature(
        hdiutil_errno=5,
        container_visible=False,
        shadow_attach_outcome=ShadowOutcome.not_attempted(),
    )
    print(sig.to_dict())
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Optional


# ---------------------------------------------------------------------------
# ShadowOutcome — tri-state value type
# ---------------------------------------------------------------------------


@dataclass(frozen=True, slots=True)
class ShadowOutcome:
    """Outcome of a readonly shadow-attach attempt.

    Construct via the three factory classmethods; do not call ``__init__``
    directly.

    Attributes:
        kind: One of ``"not_attempted"``, ``"success"``, ``"failure"``.
        fsck_clean: When ``kind="success"``, True if ``fsck_apfs -yn`` ran
            without errors. None otherwise.
        reason: When ``kind="failure"``, the error string from stderr. None
            otherwise.
    """

    kind: str
    fsck_clean: Optional[bool]
    reason: Optional[str]

    # -- Factories -----------------------------------------------------------

    @classmethod
    def not_attempted(cls) -> ShadowOutcome:
        """Primary attach succeeded, or Class A pre-flight short-circuited.

        Returns:
            A ShadowOutcome with kind ``"not_attempted"``.
        """
        return cls(kind="not_attempted", fsck_clean=None, reason=None)

    @classmethod
    def success(cls, *, fsck_clean: bool) -> ShadowOutcome:
        """Shadow attach succeeded; fsck result carried.

        Args:
            fsck_clean: True if ``fsck_apfs -yn`` found no errors on the
                shadow slice.

        Returns:
            A ShadowOutcome with kind ``"success"``.
        """
        return cls(kind="success", fsck_clean=fsck_clean, reason=None)

    @classmethod
    def failure(cls, *, reason: str) -> ShadowOutcome:
        """Shadow attach itself failed.

        Args:
            reason: stderr excerpt or short description of why the shadow
                attach failed.

        Returns:
            A ShadowOutcome with kind ``"failure"``.
        """
        return cls(kind="failure", fsck_clean=None, reason=reason)

    # -- Serialisation -------------------------------------------------------

    def to_dict(self) -> dict[str, Any]:
        """Return the Phoenix wire-format dict for this outcome.

        The schema matches ``evidence.shadow_attach_outcome`` in
        ``POST /api/devdrive/backends/{backend_id}/state`` (LFG-98 §4.4).

        Returns:
            A dict with a ``kind`` key and, conditionally, ``fsck_clean``
            or ``reason``.
        """
        result: dict[str, Any] = {"kind": self.kind}
        if self.kind == "success" and self.fsck_clean is not None:
            result["fsck_clean"] = self.fsck_clean
        if self.kind == "failure" and self.reason is not None:
            result["reason"] = self.reason
        return result

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> ShadowOutcome:
        """Reconstruct a ShadowOutcome from a wire-format dict.

        Args:
            data: Dict with at least a ``kind`` key.

        Returns:
            A ShadowOutcome instance.

        Raises:
            ValueError: When ``kind`` is not a recognised value.
        """
        kind = data.get("kind", "")
        if kind == "not_attempted":
            return cls.not_attempted()
        if kind == "success":
            return cls.success(fsck_clean=bool(data.get("fsck_clean", False)))
        if kind == "failure":
            return cls.failure(reason=str(data.get("reason", "")))
        raise ValueError(f"Unknown ShadowOutcome kind: {kind!r}")


# ---------------------------------------------------------------------------
# CorruptionSignature
# ---------------------------------------------------------------------------


@dataclass(frozen=True, slots=True)
class CorruptionSignature:
    """3-tuple canonical classification input for DevDrive failure-class routing.

    Mirrors the Swift ``CorruptionSignature`` struct (LFG-99 §1) and the JSON
    evidence schema for ``POST /api/devdrive/backends/{backend_id}/state``
    (LFG-98 §4.4, ADR-003).

    All three language runtimes (Swift, Python, Phoenix) produce and consume
    this exact tuple; Phoenix stores it verbatim as ``LifecycleEvent.evidence``.

    Attributes:
        hdiutil_errno: The errno value from the ``hdiutil attach`` exit (e.g.
            ``5`` for ``EIO``/Resource busy). ``None`` when the attach
            produced no numeric errno (image not yet probed, or Class A
            pre-flight short-circuits before any attach attempt).
        container_visible: True when ``hdiutil info`` shows the image already
            attached to a device node. False when no device node is present.
        shadow_attach_outcome: Result of a readonly shadow-attach probe
            (``hdiutil attach -readonly -shadow … -nomount``). Use
            ``ShadowOutcome.not_attempted()`` when no shadow was needed.

    Examples:
        Class A (stale half-attach — autonomous reclaimable)::

            CorruptionSignature(
                hdiutil_errno=None,
                container_visible=True,
                shadow_attach_outcome=ShadowOutcome.not_attempted(),
            )

        Class B (superblock corruption — route to recovery agent)::

            CorruptionSignature(
                hdiutil_errno=5,
                container_visible=False,
                shadow_attach_outcome=ShadowOutcome.success(fsck_clean=False),
            )
    """

    hdiutil_errno: Optional[int]
    container_visible: bool
    shadow_attach_outcome: ShadowOutcome

    # -- Serialisation -------------------------------------------------------

    def to_dict(self) -> dict[str, Any]:
        """Serialise to the Phoenix evidence wire format (LFG-98 §4.4).

        Returns:
            A dict with keys ``hdiutil_errno``, ``container_visible``, and
            ``shadow_attach_outcome`` ready for embedding in a lifecycle
            transition POST body.
        """
        return {
            "hdiutil_errno": self.hdiutil_errno,
            "container_visible": self.container_visible,
            "shadow_attach_outcome": self.shadow_attach_outcome.to_dict(),
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> CorruptionSignature:
        """Reconstruct a CorruptionSignature from a wire-format dict.

        Args:
            data: Dict with keys matching ``to_dict`` output.

        Returns:
            A CorruptionSignature instance.
        """
        errno_raw = data.get("hdiutil_errno")
        shadow_raw = data.get("shadow_attach_outcome", {"kind": "not_attempted"})
        return cls(
            hdiutil_errno=int(errno_raw) if errno_raw is not None else None,
            container_visible=bool(data.get("container_visible", False)),
            shadow_attach_outcome=ShadowOutcome.from_dict(shadow_raw),
        )
